### PR TITLE
HTML5 Local Storage, Safari Support, Time Sync

### PR DESCRIPTION
**Pull request by sughodke**
Safari was a browser that was having issues with the Client Window logging mechanism that we initially had. With this PR I've now added Local Storage logging, and the supporting runner changes.

The Timestamp synchronization was an attempt at working around the SauceLabs timing stuff, its still work in progress.

Let me know which commits to squash.  The `Mocha Exploded on Travis` commit is backing out the changes in `Using appropriate flag for Travis check`.

### CODE REVIEW

**Comment by jaridmargolin** [comment]
On initial scan, it looks like you left the window logger in. Lets just rip all of that code out. LocalStorage has support going back to IE8: http://caniuse.com/#search=localstorage

**Comment by sughodke** [comment]
Correct I left it in there I guess sentimentally 🙂.   Then it became more of a hassle to keep it supported. Now it's dead code. Will rip it all out.

**Comment by sughodke** [comment]
Removed Window Logger.  Note that I can't refactor `dumpLogger` function to live inside the `createLocalStorageLogger` object.  As that object is only instantiated when the RcSocket is created.  In certain tests we need to access the logs without instantiating the RcSocket.

Should be noted, that we don't really need to tie the Logger creation to Socket creation.  This was carried over from the previous implementation.

File: test/cross-browser/reference.html [hunk 1]
```diff
@@ -22,36 +20,51 @@
...
         };
 
+        var dumpLogger = function() {
+            var buffer = [];
+            for (var i = 0; i < localStorage.length; i++) {
+                buffer.push({
+                    ts: parseInt(localStorage.key(i)),
+                    message: localStorage.getItem(localStorage.key(i))
+                });
+            }
```
Comment (jaridmargolin) [inline test/cross-browser/reference.html]: I lightly tested this inside of a chrome console and it worked. Can't speak for other browsers.

```
for (var key in localStorage) {
}
```

File: test/cross-browser/reference.html [hunk 2]
```diff
@@ -22,36 +20,51 @@
 
     <!-- traces -->
     <script>
-        var childWin;
+        var clientLogger;
+
+        function createLocalStorageLogger() {
+            return {
+                debug: function() {
+                    localStorage.setItem(
```
Comment (jaridmargolin) [inline test/cross-browser/reference.html]: This is a blanket comment. The majority of our implemented code, accesses localStorage using bracket notation rather than `setItem` and `getItem`. It appears both have near identical behavior: http://stackoverflow.com/questions/12632463/is-localstorage-getitemitem-better-than-localstorage-item-or-localstoragei?answertab=active#tab-top

If it were me I would change to use bracket notation (I prefer consistency), but ultimately your call.

File: test/cross-browser/report.js [hunk 3]
```diff
@@ -29,14 +30,23 @@ var clearLogger = function() {
...
-        return logger.push(z);
+    var timestampPromise = Q.fcall(function() {
+        return Date.now() + tsOffset;
     });
+
+    //if (typeof driver !== 'undefined') {
+    //    timestampPromise = driver.executeScript('return Date.now();');
+    //}
+
+    return timestampPromise
```
Comment (jaridmargolin) [inline test/cross-browser/report.js]: Does this work? Its not invoked.

File: test/cross-browser/report.js [hunk 4]
```diff
@@ -29,14 +30,23 @@ var clearLogger = function() {
...
+        .then(function(d) {
+            var z = {
+                ts: d,
+                message: msg
+            };
+
+            return logger.push(z);
```
Comment (jaridmargolin) [inline test/cross-browser/report.js]: As z is not used anywhere else, I would just inline it inside of push.

File: test/cross-browser/report.js [hunk 5]
```diff
@@ -141,29 +160,40 @@ var createRcSocket = function() {
...
-                    console.log(d + service + ': ' + m.join(' '));
-                });
-            })
-            .then(getParentWindow);
-    });
+    return driver
+        .executeScript('return window.dumpLogger();')
+        .then(function(z) {
+            var allLogs = logger.concat(z);
+            allLogs.sort(function(a, b) {
```
Comment (jaridmargolin) [inline test/cross-browser/report.js]: This is embarrassing, but I legitimately have never used the native `Array.sort` method. I didn't know it existed? Or let me rephrase, I thought it was an ES5 method like `map`... Turns out its been in the spec since day 1.

File: test/cross-browser/report.js [hunk 6]
```diff
@@ -141,29 +160,40 @@ var createRcSocket = function() {
...
+                return a.ts - b.ts;
+            });
+
+            allLogs.map(function(log) {
```
Comment (jaridmargolin) [inline test/cross-browser/report.js]: Not 100% sure exactly whats happening in here, but assuming it is all just required formatting.

File: test/cross-browser/report.js [hunk 7]
```diff
@@ -270,8 +308,10 @@ describe('RcSocketIntegration', function () {
             .then(waitASecond)
             .then(stopServerSocket)
             .then(waitASecond)
+            .then(waitASecond)
             .then(startServerSocket)
             .then(waitASecond)
+            .then(waitASecond)
```
Comment (jaridmargolin) [inline test/cross-browser/report.js]: `waitTwoSeconds`? 😂 Haha sorry, I am just realizing how gritty and hacky this project was.

To cleaner and less hacky code in the new year 🍻

**Comment by jaridmargolin** [comment]
Left a few comments. Also going to open an issue to rework clientLogger a little bit.

**Comment by jaridmargolin** [comment]
Also just a little bit of cruft left in from our window logging solution `test/cross-browser/report.js` (106 - 137).

Btw, sorry. The window logging seemed like a good idea at the time you implemented 😒

File: test/cross-browser/reference.html (hunk 2 above)
Comment (sughodke) [inline test/cross-browser/reference.html]: I'm leaning more towards the `setItem`/`getItem` implementation.  I read through the SO article you mentioned, I feel like localStorage is not a traditional array object (as mentioned by one answer, length is not settable).  So I feel that it's wise not to treat it like one... Yeah?

File: test/cross-browser/reference.html (hunk 1 above)
Comment (sughodke) [inline test/cross-browser/reference.html]: See comment below, and also in this SO article, folks get inconsistent results from the forEach way of reading localStorage.

http://stackoverflow.com/a/3138591

File: test/cross-browser/report.js (hunk 3 above)
Comment (sughodke) [inline test/cross-browser/report.js]: It is invoked by the caller of this function, the `then` clause triggers the promise to evaluate itself.  

I will remove the commented out code, this was an early implementation of the timestamp sync.

File: test/cross-browser/report.js (hunk 4 above)
Comment (sughodke) [inline test/cross-browser/report.js]: This function has been cleaned up

File: test/cross-browser/report.js (hunk 6 above)
Comment (sughodke) [inline test/cross-browser/report.js]: Yes thats correct.  Just checked the documentation on `map`, it will apply the function to all the elements of the array **in order**.

map calls a provided callback function once for each element in an array, in order, ...

**Comment by sughodke** [comment]
> Also just a little bit of cruft left in from our window logging solution test/cross-browser/report.js (106 - 137).

Thanks cleaned up.

> Btw, sorry. The window logging seemed like a good idea at the time you implemented 😒

I think it was a great solution at the time.  Now, if you had said we need to load a Flash plugin which has a custom Event handler that persists log messages.  Then things might be different.  :dancer:

File: test/cross-browser/reference.html (hunk 2 above)
Comment (jaridmargolin) [inline test/cross-browser/reference.html]: My initial thought was, "who cares? why would I ever want to set length, key, setItem, getItem, removeItem, or clear?" It would be horrible practice to do so.

But I buy the argument of using function implementation so that it is apparent it has different behavior than default objects.

File: test/cross-browser/reference.html (hunk 1 above)
Comment (jaridmargolin) [inline test/cross-browser/reference.html]: I'd be curious to see the browser tests on this but :+1:
